stable/openebs/values.yaml

```diff
@@ -104,11 +104,14 @@ ndm:
   healthCheck:
     initialDelaySeconds: 30
     periodSeconds: 60
+  featureGates:
```

Comment:
@akhilerm -- can you check if there are any best practices around how to specify the featurettes. 

A use case that I am thinking about is, let us say there is an alpha feature that needs to be provided, but it may impact multiple components. For example, I would like to create a feature gate for cStor new schema, which will have to include new deployments as well as pass the flag to m-apiserver to take some corrective  actions. 

Is it then better to have, featureGates at the top level:
```
featuerGates:
  enabled: false
  ndm:
    features: "GPTBasedUUID"
  cstor:
    features: "CSIDriver"
```

or

```
featuerGates:
  ndmGPTBasedUUID:
    enabled: false
  cStorCSIDriver
    enabled: false
```

In the second approach, probably within the template handle the filling of the required fields. May be there is a template function that can be written to generate a string per component based not he helm values, like the serviceAccount Template functions.

Reply:
```
featuerGates:
  ndm:
    enabled: true
    features: "GPTBasedUUID,FeatureABC"
  cstor:
    enabled: false
    features: "CSIDriver"
```
Slight modification to the first one, looks good to me.

Reply:
only one drawback with this comma separate approach is that, it may not allow for templating changes. Maybe there is a solution. For example if ndm.featureABC is enabled, if we need to include additional configuration to daemonset or operator pod etc. Also there could be some additional configurations that we may have to pass for certain features to be enabled.

So the featureGates should allow us to:
- Enable/disable a feature that crosses component boundaries.
- Modify the object deployed depending on the feature enabled. 
- Take in additional configuration to enable the feature.

Reply:
@kmova, This is an approach for features and feature gates. A feature is something that is available across all of openebs. It can be a featuregate or something else like setting an env.
This allows multiple components to use the same feature.

```
# this is specific for feature gates for components. The individual features are available inside features
featureGates:
  ndm:
    enabled: true

# this is the list of features available across all the components. Any component can use them
features:
  GPTBasedUUID:
    enabled: true
    featureGateFlag: "GPTBasedUUID"
  LegacycStor:
    enabled: false
  FeatureABC:
    enabled: false
    featureGateFlag: "FeatureABC"
```

This is how the feature gate will be set by NDM:
```
{{- if .Values.featureGates.ndm }}
          - --feature-gates={{ .Values.features.GPTBasedUUID.featureGateFlag }}={{ .Values.features.GPTBasedUUID.enabled}}
          - --feature-gates={{ .Values.features.FeatureABC.featureGateFlag }}={{ .Values.features.FeatureABC.enabled}}
{{- end}}
```

Reply:
The `features` section is good.  Just check if this will help: 

```
# this is the list of features available across all the components. Any component can use them
featureGates:
  enabled: true
  GPTBasedUUID:
    enabled: true
    flag: "GPTBasedUUID"
  LegacycStor:
    enabled: false
  FeatureABC:
    enabled: false
    flag: "FeatureABC"
```

Then use as follow:
```
{{- if .Values.featureGates.enabled }}
{{- if .Values.featureGates.GPTBasedUUID.enabled }}
          - --feature-gates={{ .Values.featureGates.GPTBasedUUID.flag }}={{ .Values.featuresGates.GPTBasedUUID.enabled}}
{{- end}}
{{- if .Values.featureGates.FeatureABC.enabled }}
          - --feature-gates={{ .Values.featureGates.FeatureABC.flag }}={{ .Values.featureGates.FeatureABC.enabled}}
{{- end}}
{{- end}}
```

Reply:
With this approach we can't have per component feature gate / features.
`Features` can be feature gates or something like env or something else.

In the approach you mentioned, if `featureGates.enabled` is true, then for all components it will be enabled. The reason I separated `features` and `featureGates` was to get per component control.

Reply:
Initially like the idea of per component, then thinking about it a bit further realized: 
- we have close to 10/15 components if we take individual containers, and atleast 6+ components based on dataplane and control plane.
- the individual feature flags should suffice to enable/disable across the components + a global flag that says this installation uses feature gates and is not production-ready is a good way to keep the messaging short.

Reply:
Btw, why is the list-test failing? cc: @prateekpandey14

Reply:
@kmova . I think at the time of raising this PR, all the RC1 images were not available.

Reply:
> a global flag that says this installation uses feature gates and is not production-ready is a good way to keep the messaging short.

Cool , will use that approach and update.